### PR TITLE
feat:  Add new Remap interface to relations

### PR DESCRIPTION
**Pull request by EpsilonPrime**
Remap() is an alternative to the current Remap variants currently found on the Builder interface.  It preserves the existing behavior including checking the proposed mapping for validity.  It also modifies the relation instead of returning a copy.  All of the existing Remap function variants have been marked as deprecated in favor of Remap.

### CODE REVIEW

File: plan/relations.go [hunk 1]
```diff
@@ -581,6 +593,18 @@ func (p *ProjectRel) CopyWithExpressionRewrite(rewriteFunc RewriteFunc, newInput
 	return &proj, nil
 }
 
+func (p *ProjectRel) ChangeMapping(mapping []int32) error {
+	noutput := int32(len(p.Remap(p.RecordType()).Types))
```
Comment (jacques-n) [inline plan/relations.go:597]: If this operation is already mapped, we should be applying the mapping on top of the existing one, not replacing it.

File: plan/plan.go [hunk 2]
```diff
@@ -275,6 +275,7 @@ type Rel interface {
 	// record type from the output RecordType of the underlying Relation.
 	// If the output mapping is nil, then this just returns the input
 	// type.
+	// TODO - Rename this ApplyMap.
```
Comment (jacques-n) [inline plan/plan.go:278]: Add api to interface.

Also, I think we should be copying, not overwriting. Overwriting/mutating nodes in a plan can be dangerous given the possibility of duplicated subplans, etc.

Maybe call CopyWithRemap() or similar.

File: plan/relations.go [hunk 3]
```diff
@@ -148,6 +148,18 @@ func (b *baseReadRel) updateFilters(filters []expr.Expression) {
 	b.filter, b.bestEffortFilter = filters[0], filters[1]
 }
 
+func (b *baseReadRel) ChangeMapping(mapping []int32) error {
+	noutput := int32(len(b.Remap(b.RecordType()).Types))
```
Comment (jacques-n) [inline plan/relations.go:152]: We should be able to write this logic once, not many times.

File: plan/plan.go (hunk 2 above)
Comment (EpsilonPrime) [reply plan/plan.go:278]: Remap doesn't really make sense on write relations so I specifically created an alternative interface.

I'm less inclined to worry about copying as these are more of a build time artifact (i.e. not protos).  If one wants to duplicate a subplan that should do the copying.

File: plan/relations.go (hunk 1 above)
Comment (EpsilonPrime) [reply plan/relations.go:597]: I don't see any utility to remap something that's already been mapped.  I could add a check to return an error if someone attempts to remap.

File: plan/relations.go [hunk 4]
```diff
@@ -741,6 +776,12 @@ func (j *JoinRel) CopyWithExpressionRewrite(rewriteFunc RewriteFunc, newInputs .
 	return &join, nil
 }
 
+func (j *JoinRel) ChangeMapping(mapping []int32) error {
+	newMapping, err := ChangeMapping(j, mapping)
+	j.mapping = newMapping
```
Comment (jacques-n) [inline plan/relations.go:781]: If there is an error, we lose the old mapping if I reading this right. If err, I would expect no change.

File: plan/plan.go [hunk 5]
```diff
@@ -271,6 +271,16 @@ type Rel interface {
...
+	// ClearMapping resets the mapping for this relation.
+	ClearMapping()
+	// ChangeMapping modifies the current relation by applying the provided
+	// mapping to the current relation.  Typically used to remove any unneeded
+	// columns or provide them in a different order.  If there already is a
+	// mapping on this relation, this provides mapping over the current mapping.
+	//
+	// If any column numbers specified are outside the currently available input
+	// range an error is returned and the mapping is left unchanged.
+	ChangeMapping(mapping []int32) error
```
Comment (jacques-n) [inline plan/plan.go:283]: Can we have this make a copy as opposed to an inplace edit? 

I'm also more OverlayMapping or ApplyMapping as a name.

Also, would be good to use vararg.

File: plan/plan.go [hunk 6]
```diff
@@ -271,6 +271,16 @@ type Rel interface {
 	// result should be 3 columns consisting of the 5th, 2nd and 1st
 	// output columns from the underlying relation.
 	OutputMapping() []int32
+	// ClearMapping resets the mapping for this relation.
+	ClearMapping()
```
Comment (jacques-n) [inline plan/plan.go:275]: I don't think we should add this.

File: plan/plan.go [hunk 7]
```diff
@@ -271,6 +271,16 @@ type Rel interface {
 	// result should be 3 columns consisting of the 5th, 2nd and 1st
 	// output columns from the underlying relation.
 	OutputMapping() []int32
```
Comment (jacques-n) [inline plan/plan.go:273]: I think we should deprecate this.

Comment (EpsilonPrime) [reply plan/plan.go:273]: done.

File: plan/relations.go (hunk 4 above)
Comment (EpsilonPrime) [reply plan/relations.go:783]: Fixed the bug in the ChangeMapping helper and added a test.

File: plan/builders.go [hunk 8]
```diff
@@ -71,44 +71,60 @@ type Builder interface {
...
+	// Deprecated: Use Fetch(...).ChangeMapping() instead.
 	JoinAndFilter(left, right Rel, condition, postJoinFilter expr.Expression, joinType JoinType) (*JoinRel, error)
+	// Deprecated: Use Join(...).ChangeMapping() instead.
 	JoinRemap(left, right Rel, condition expr.Expression, joinType JoinType, remap []int32) (*JoinRel, error)
 	Join(left, right Rel, condition expr.Expression, joinType JoinType) (*JoinRel, error)
+	// Deprecated: Use NamedScan(...).ChangeMapping() instead.
 	NamedScanRemap(tableName []string, schema types.NamedStruct, remap []int32) (*NamedTableReadRel, error)
 	NamedScan(tableName []string, schema types.NamedStruct) *NamedTableReadRel
+	// Deprecated: Use NamedWriteMap(...).ChangeMapping() instead.
 	NamedWriteRemap(input Rel, op WriteOp, tableName []string, schema types.NamedStruct, remap []int32) (*NamedTableWriteRel, error)
```
Comment (jacques-n) [inline plan/builders.go:114]: Looks like deprecation message is wrong and NamedWrite() is missing.

File: plan/plan.go (hunk 6 above)
Comment (EpsilonPrime) [reply plan/plan.go]: Removed.

File: plan/builders.go (hunk 8 above)
Comment (EpsilonPrime) [reply plan/builders.go:114]: fixed.

File: plan/plan.go (hunk 5 above)
Comment (EpsilonPrime) [reply plan/plan.go]: Renamed to `Remap()`.  There could be some confusion with `remap()` which we could rename to `applyMapping()`.

File: plan/relations.go [hunk 9]
```diff
@@ -212,6 +240,13 @@ func (n *NamedTableReadRel) CopyWithExpressionRewrite(rewriteFunc RewriteFunc, _
 	return &nt, nil
 }
 
+func (n *NamedTableReadRel) Remap(mapping ...int32) (Rel, error) {
+	newMapping, err := RemapHelper(n, mapping)
+	newRel := n
```
Comment (jacques-n) [inline plan/relations.go:245]: Since this is a pointer receiver, I believe this isn't a copy, right?

File: plan/relations.go [hunk 10]
```diff
@@ -152,6 +152,34 @@ func (b *baseReadRel) updateFilters(filters []expr.Expression) {
 	b.filter, b.bestEffortFilter = filters[0], filters[1]
 }
 
+type MappableRel interface {
+	RecordType() types.RecordType
+	OutputMapping() []int32
+}
+
+// RemapHelper implements the core functionality of RemapHelper for relations.
+// It returns the relation's existing mapping on an error to ease being called.
```
Comment (jacques-n) [inline plan/relations.go:161]:
```suggestion
// It returns the relation's existing mapping or an error to ease being called.
```

File: plan/relations.go [hunk 11]
```diff
@@ -267,6 +302,13 @@ func (v *VirtualTableReadRel) CopyWithExpressionRewrite(rewriteFunc RewriteFunc,
 	return &vtr, nil
 }
 
+func (v *VirtualTableReadRel) Remap(mapping ...int32) (Rel, error) {
+	newMapping, err := RemapHelper(v, mapping)
```
Comment (jacques-n) [inline plan/relations.go:306]: If I were you, I'd introduce a non exported set mapping function on relation and then this code could be factored out to a single invocation.


something like 
```
RemapHelper(r Rel, newMapping) (Rel, error) {
  newRel := r.Copy(v.GetInputs())
  err := r.setMapping(newMapping)
  return newRel, err
}
```

File: plan/relations.go (hunk 9 above)
Comment (EpsilonPrime) [reply plan/relations.go:245]: fixed.

**Review by jacques-n — APPROVED** [review]

File: plan/relations.go (hunk 11 above)
Comment (EpsilonPrime) [reply plan/relations.go]: Adding the unexported method to the Rel interface was the trick I needed to get that to work.  Thanks!

**Pull request merged**